bluemira/equilibria/harmonics.py

```diff
@@ -0,0 +1,682 @@
+    )
+
+    # Set min to save some time
+    min_degree = 2
+    max_degree = len(collocation["x"]) - 1
+
+    for degree in np.arange(min_degree, max_degree + 1):
+        # Construct matrix from harmonic amplitudes for coils
+        currents2harmonics = coil_harmonic_amplitude_matrix(
+            sh_coilset.get_control_coils(), degree, r_t
+        )
+
+        # Calculate necessary coil currents
+        currents, residual, rank, s = np.linalg.lstsq(
+            currents2harmonics, psi_harmonic_ampltidues[:degree], rcond=None
+        )
+
+        # Calculate the coilset SH amplitudes for use in optimisisation
+        coil_current_harmonic_ampltidues = currents2harmonics @ currents
+
+        # Set currents in coilset
+        sh_coilset.get_control_coils().current = currents
+
+        # Calculate the approximate Psi contribution from the coils
+        coilset_approx_psi = sh_coilset.psi(grid.x, grid.z)
+
+        # Total
+        approx_total_psi = coilset_approx_psi + plasma_psi
+
+        # Get plasma boundary for comparison to starting equilibrium using fit metric
+        eq_copy.coilset = sh_coilset
+        approx_LCFS = eq_copy.get_LCFS(psi=approx_total_psi)
+
+        # Compare staring equlibrium to new approximate equilibrium
+        fit_metric_value = lcfs_fit_metric(original_LCFS, approx_LCFS)
+
+        if fit_metric_value <= acceptable_fit_metric:
+            bluemira_print(
+                f"The fit metric value acheived is {fit_metric_value} using {degree} degrees."
+            )
+            break
+        elif degree == max_degree:
+            raise BluemiraError(
+                f"Uh oh, you may need to use more degrees for a fit metric of {acceptable_fit_metric}! Use a greater number of collocation points please."
+            )
+
+    # plot comparing orginal psi to the SH approximation
+    if plot:
+        plot_psi_comparision(
+            grid,
+            eq.psi(grid.x, grid.z),
+            approx_total_psi,
+            eq.coilset.psi(grid.x, grid.z),
+            coilset_approx_psi,
+        )
+
+    return (
+        sh_coilset,
+        r_t,
+        coil_current_harmonic_ampltidues,
+        degree,
+        fit_metric_value,
+        approx_total_psi,
+    )
+
+
+def plot_psi_comparision(
+    grid: Grid,
+    tot_psi_org: np.ndarray,
+    tot_psi_app: np.ndarray,
+    vac_psi_org: np.ndarray,
+    vac_psi_app: np.ndarray,
+):
+    """
+    Create plot comparing an orginal psi to psi obtained from approximation.
+
+    Parameters
+    ----------
+    grid: Bluemira Grid
+        Nedd x and z values to plot psi.
+    tot_psi_org: ndarray
+        Original Total Psi
+    tot_psi_app: ndarray
+        Approximation Total Psi
+    vac_psi_org: ndarray
+        Original Vacuum Psi (contribution from entire coilset)
+    vac_psi_app: ndarray
+        Approximation Vacuum Psi (contribution from entire coilset)
+
+    """
+    nlevels = 50  # PLOT_DEFAULTS["psi"]["nlevels"]
```

Comment:
can we make this an optional arg with a default of 50. I dont know why I'd want to change it but its nice to have the option. Unless there is a reason for it?